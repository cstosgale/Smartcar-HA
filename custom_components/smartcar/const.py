from enum import StrEnum, auto

DOMAIN = "smartcar"
DEFAULT_NAME = "Smartcar"
API_HOST = "https://api.smartcar.com"

PLATFORMS = [
    "sensor",
    "switch",
    "lock",
    "device_tracker",
    "binary_sensor",
    "number",
]

OAUTH2_AUTHORIZE = "https://connect.smartcar.com/oauth/authorize"
OAUTH2_TOKEN = "https://auth.smartcar.com/oauth/token"
SMARTCAR_MODE = "live"


class Scope(StrEnum):
    READ_VEHICLE_INFO = auto()
    READ_VIN = auto()
    READ_BATTERY = auto()
    READ_CHARGE = auto()
    READ_ENGINE_OIL = auto()
    READ_FUEL = auto()
    READ_LOCATION = auto()
    READ_ODOMETER = auto()
    READ_SECURITY = auto()
    READ_TIRES = auto()
    CONTROL_CHARGE = auto()
    CONTROL_SECURITY = auto()


REQUIRED_SCOPES = [
    Scope.READ_VEHICLE_INFO,
    Scope.READ_VIN,
]

CONFIGURABLE_SCOPES = [scope for scope in Scope if scope not in REQUIRED_SCOPES]

DEFAULT_SCOPES = [
    Scope.READ_BATTERY,
    Scope.READ_CHARGE,
    Scope.READ_LOCATION,
    Scope.READ_ODOMETER,
    Scope.READ_SECURITY,
    Scope.READ_VEHICLE_INFO,
    Scope.READ_VIN,
    Scope.CONTROL_CHARGE,
]
