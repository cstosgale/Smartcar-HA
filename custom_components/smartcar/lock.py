from __future__ import annotations

import logging
from aiohttp import ClientResponseError
from homeassistant.exceptions import ConfigEntryAuthFailed, HomeAssistantError
from homeassistant.components.lock import LockEntity, LockEntityDescription
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from .const import DOMAIN, API_BASE_URL_V1  # Uses V1 for POST
from .coordinator import SmartcarVehicleCoordinator, SmartcarCoordinatorEntity

_LOGGER = logging.getLogger(__name__)
ENTITY_DESCRIPTIONS: tuple[LockEntityDescription, ...] = (
    LockEntityDescription(key="door_lock", name="Door Lock"),
)


async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
) -> None:
    coordinators: dict[str, SmartcarVehicleCoordinator] = (
        entry.runtime_data.coordinators
    )
    session = entry.runtime_data.session
    entities = []
    for vin, coordinator in coordinators.items():
        for description in ENTITY_DESCRIPTIONS:
            if coordinator.is_scope_enabled(description.key, verbose=True):
                entities.append(
                    SmartcarDoorLock(coordinator, session, entry, description)
                )
    _LOGGER.info("Adding %d Smartcar lock entities", len(entities))
    async_add_entities(entities)


class SmartcarDoorLock(SmartcarCoordinatorEntity, LockEntity):
    _attr_has_entity_name = True

    def __init__(self, coord, session, entry, desc):
        super().__init__(coord, desc)
        self.vin = coord.vin
        self.vehicle_id = coord.vehicle_id
        self.session = session
        self.entry = entry
        self.entity_description = desc
        self._attr_unique_id = f"{self.vin}_{desc.key}"
        self._attr_device_info = {"identifiers": {(DOMAIN, self.vin)}}

    @property
    def is_locked(self):
        data = self.coordinator.data
        lock_data = data.get("lock_status") if data else None
        return lock_data.get("isLocked") if lock_data else None

    async def _async_send_lock_command(self, action):  # Error handling kept
        _LOGGER.info("Attempting to %s doors for %s", action, self.vin)
        url = f"{API_BASE_URL_V1}/vehicles/{self.vehicle_id}/security"
        payload = {"action": action}
        try:
            resp = await self.session.async_request("post", url, json=payload)
            resp.raise_for_status()
            _LOGGER.info(
                "%s cmd OK for %s", action.capitalize(), self.vin
            )  # Removed early refresh
        except ClientResponseError as e:
            if e.status in (400, 401, 403):
                _LOGGER.warning(
                    "Auth err [%s] sending %s cmd for %s", e.status, action, self.vin
                )
                self.entry.async_start_reauth(self.hass)
            else:
                _LOGGER.error("HTTP err sending %s cmd for %s: %s", action, self.vin, e)
                raise HomeAssistantError(f"API err {e.status}") from e
        except ConfigEntryAuthFailed:
            _LOGGER.warning("AuthFail sending %s cmd for %s", action, self.vin)
            self.entry.async_start_reauth(self.hass)
        except Exception as e:
            _LOGGER.exception("Unexpected err sending %s cmd for %s", action, self.vin)
            raise HomeAssistantError("Unexpected lock err") from e

    async def async_lock(self, **kwargs):
        await self._async_send_lock_command("LOCK")

    async def async_unlock(self, **kwargs):
        await self._async_send_lock_command("UNLOCK")

    @property
    def available(self):
        return (
            super().available
            and self.coordinator.data is not None
            and self.coordinator.data.get("lock_status") is not None
            and "isLocked" in self.coordinator.data["lock_status"]
        )
