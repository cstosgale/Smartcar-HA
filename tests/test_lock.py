"""Test lock entities."""

from unittest.mock import AsyncMock

from aiohttp import ClientResponseError
from homeassistant.components.lock import DOMAIN as LOCK_DOMAIN, LockState
from homeassistant.const import ATTR_ENTITY_ID, SERVICE_LOCK, SERVICE_UNLOCK
from homeassistant.core import HomeAssistant
import pytest
from pytest_homeassistant_custom_component.common import MockConfigEntry
from pytest_homeassistant_custom_component.test_util.aiohttp import AiohttpClientMocker
from syrupy.assertion import SnapshotAssertion

from . import MOCK_API_ENDPOINT, setup_integration

NO_ERROR = None.__class__


@pytest.mark.parametrize(
    (
        "service_action",
        "api_status",
        "api_status_slug",
        "expected_state",
        "expected_raises",
    ),
    [
        (SERVICE_LOCK, 200, "success", LockState.LOCKED, NO_ERROR),
        (SERVICE_UNLOCK, 200, "success", LockState.UNLOCKED, NO_ERROR),
        (SERVICE_LOCK, 409, "unreachable", LockState.UNLOCKED, NO_ERROR),
        (SERVICE_UNLOCK, 409, "unreachable", LockState.UNLOCKED, NO_ERROR),
        (SERVICE_LOCK, 401, "unauthroized", LockState.UNLOCKED, NO_ERROR),
        (SERVICE_UNLOCK, 401, "unauthroized", LockState.UNLOCKED, NO_ERROR),
        (SERVICE_UNLOCK, 500, "server", LockState.UNLOCKED, ClientResponseError),
    ],
)
@pytest.mark.parametrize("vehicle_fixture", ["unknown_make"])
async def test_lock(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    aioclient_mock: AiohttpClientMocker,
    snapshot: SnapshotAssertion,
    vehicle: AsyncMock,
    service_action: str,
    api_status: int,
    api_status_slug: str,
    expected_state: LockState,
    expected_raises: Exception,
) -> None:
    """Test locking doors."""

    await setup_integration(hass, mock_config_entry)
    assert len(aioclient_mock.mock_calls) == 1

    aioclient_mock.post(
        f"{MOCK_API_ENDPOINT}/v2.0/vehicles/{vehicle['id']}/security",
        status=api_status,
        json={
            "message": "Some message related to the action unused by our code",
            "status": api_status_slug,
        },
    )

    try:
        await hass.services.async_call(
            LOCK_DOMAIN,
            service_action,
            {ATTR_ENTITY_ID: "lock.smartcar_784n_door_lock"},
            blocking=True,
        )
    except Exception as error:  # noqa: BLE001
        raised_error = error
    else:
        raised_error = None  # type: ignore[assignment]

    lock_state = hass.states.get("lock.smartcar_784n_door_lock")
    assert lock_state.state == expected_state
    assert isinstance(
        raised_error,
        expected_raises,  # type: ignore[arg-type]
    )

    assert len(aioclient_mock.mock_calls) == 2
    assert [tuple(mock_call) for mock_call in aioclient_mock.mock_calls[1:]] == snapshot
