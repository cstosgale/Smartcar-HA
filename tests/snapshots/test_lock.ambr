# serializer version: 1
# name: test_lock[unknown_make-lock-200-success-locked-NoneType]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/security'),
      dict({
        'action': 'LOCK',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_lock[unknown_make-lock-401-unauthroized-unlocked-NoneType]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/security'),
      dict({
        'action': 'LOCK',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_lock[unknown_make-lock-409-unreachable-unlocked-NoneType]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/security'),
      dict({
        'action': 'LOCK',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_lock[unknown_make-unlock-200-success-unlocked-NoneType]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/security'),
      dict({
        'action': 'UNLOCK',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_lock[unknown_make-unlock-401-unauthroized-unlocked-NoneType]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/security'),
      dict({
        'action': 'UNLOCK',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_lock[unknown_make-unlock-409-unreachable-unlocked-NoneType]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/security'),
      dict({
        'action': 'UNLOCK',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_lock[unknown_make-unlock-500-server-unlocked-ClientResponseError]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/security'),
      dict({
        'action': 'UNLOCK',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
